Examples/Algorithms/TrackFinding/src/HoughTransformSeeder.cpp

```diff
@@ -0,0 +1,479 @@
+    std::swap(x_min, x_max);
+  if (x_max < m_cfg.xMin || x_min > m_cfg.xMax)
+    return {0, 0};  // out of bounds
+
+  // Get bins
+  int x_bin_min = quant(m_cfg.xMin, m_cfg.xMax, m_cfg.houghHistSize_x, x_min);
+  int x_bin_max =
+      quant(m_cfg.xMin, m_cfg.xMax, m_cfg.houghHistSize_x, x_max) +
+      1;  // exclusive
+
+  // Extend bins
+  unsigned extend = getExtension(yBin_min, layer);
+  x_bin_min -= extend;
+  x_bin_max += extend;
+
+  // Clamp bins
+  if (x_bin_min < 0)
+    x_bin_min = 0;
+  if (x_bin_max > static_cast<int>(m_cfg.houghHistSize_x))
+    x_bin_max = m_cfg.houghHistSize_x;
+
+  return {x_bin_min, x_bin_max};
+}
+
+// We allow variable extension based on the size of m_hitExtend_x. See comments
+// below.
+unsigned ActsExamples::HoughTransformSeeder::getExtension(
+    unsigned y, unsigned layer) const {
+   if (m_cfg.hitExtend_x.size() == m_cfg.nLayers) {
+      return m_cfg.hitExtend_x[layer];
+   }
+
+
+  if (m_cfg.hitExtend_x.size() == m_cfg.nLayers * 2) {
+    // different extension for low pt vs high pt, split in half but irrespective
+    // of sign first nLayers entries of m_hitExtend_x is for low pt half, rest
+    // are for high pt half
+    if (y < m_cfg.houghHistSize_y / 4 || y > 3 * m_cfg.houghHistSize_y / 4)
+      return m_cfg.hitExtend_x[layer];
+    return m_cfg.hitExtend_x[m_cfg.nLayers + layer];
+  }
+  return 0;
+}
+
+/**
+ * Given a list of sizes (of arrays), generates a list of all combinations of
+ * indices to index one element from each array.
+ *
+ * For example, given [2 3], generates [(0 0) (1 0) (0 1) (1 1) (0 2) (1 2)].
+ *
+ * This basically amounts to a positional number system of where each digit has
+ * its own base. The number of digits is sizes.size(), and the base of digit i
+ * is sizes[i]. Then all combinations can be uniquely represented just by
+ * counting from [0, nCombs).
+ *
+ * For a decimal number like 1357, you get the thousands digit with n / 1000 = n
+ * / (10 * 10 * 10). So here, you get the 0th digit with n / (base_1 * base_2 *
+ * base_3);
+ */
+std::vector<std::vector<int>>
+ActsExamples::HoughTransformSeeder::getComboIndices(
+    std::vector<size_t>& sizes) const {
+  size_t nCombs = 1;
+  std::vector<size_t> nCombs_prior(sizes.size());
+  std::vector<int> temp(sizes.size(), 0);
+
+  for (size_t i = 0; i < sizes.size(); i++) {
+    if (sizes[i] > 0) {
+      nCombs_prior[i] = nCombs;
+      nCombs *= sizes[i];
+    } else
+      temp[i] = -1;
+  }
+
+  std::vector<std::vector<int>> combos(nCombs, temp);
+
+  for (size_t icomb = 0; icomb < nCombs; icomb++) {
+    size_t index = icomb;
+    for (size_t isize = sizes.size() - 1; isize < sizes.size(); isize--) {
+      if (sizes[isize] == 0)
+        continue;
+      combos[icomb][isize] = static_cast<int>(index / nCombs_prior[isize]);
+      index = index % nCombs_prior[isize];
+    }
+  }
+
+  return combos;
+}
+
+
+void ActsExamples::HoughTransformSeeder::AddSPs(const AlgorithmContext& ctx) const {
```

Comment:
Naming convention: `addSpacePoints(...)`

Reply:
Then, one could actually make this callable with a spacepoint container, like this, this could be used for full event processing or batch-wise processing.

Reply:
I changed the name, do we prefer for this to take containers? We do need a single container type then :)